18F/identity-idp
# LG-12292: Perform reCAPTCHA assessment at sign-in

## 🎫 Ticket

[LG-12292](https://cm-jira.usa.gov/browse/LG-12292)

## 🛠 Summary of changes

Implements reCAPTCHA validation at sign-in, behind a feature flag.

This builds upon code from the previous proof-of-concept at #10587.

## 📜 Testing Plan

Verify that you can pass and fail reCAPTCHA validation at sign-in from a new or existing device:

1. Go to http://localhost:3000
2. Enter email and password
3. (Optional) Customize "reCAPTCHA score" in the mock debugger tool
4. Click "Sign in"
5. Observe:
   - If you have already signed in on this device, you are allowed to proceed regardless of the score entered in Step 3
   - If you haven't already signed in on this device and you enter a score at or above 0.3, you are allowed to proceed
   - Otherwise, you are presented with an error message

Verify that there is no validation when the feature flag is disabled:

1. Add `sign_in_recaptcha_score_threshold: 0.0` and `phone_recaptcha_mock_validator: false` (rename TBD) to `config/application.yml`
2. Restart server
3. Go to http://localhost:3000
4. Observe no "reCAPTCHA score" field
5. Sign in with email and password
6. Observe no error message shown

## 👀 Screenshots

![image](https://github.com/18F/identity-idp/assets/1779930/b765bf08-ee85-4474-b96c-deafa997c0e5)
